homeassistant/components/oralb/__init__.py

```diff
@@ -1,30 +1,104 @@
 """The OralB integration."""
 from __future__ import annotations
 
+import asyncio
+from datetime import timedelta
+import logging
+
+import async_timeout
+from oralb import OralB
+
+from homeassistant.components import bluetooth
+from homeassistant.components.bluetooth.match import ADDRESS, BluetoothCallbackMatcher
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import Platform
-from homeassistant.core import HomeAssistant
+from homeassistant.const import CONF_MAC, EVENT_HOMEASSISTANT_STOP, Platform
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
-from .const import DOMAIN
+from .const import DEVICE_TIMEOUT, DOMAIN, UPDATE_SECONDS
+from .models import OralBData
 
 # For your initial PR, limit it to 1 platform.
 PLATFORMS: list[Platform] = [Platform.SENSOR]
+_LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up OralB from a config entry."""
+    address: str = entry.data[CONF_MAC]
+    print("using address: ", address)
+    ble_device = bluetooth.async_ble_device_from_address(hass, address.upper(), True)
+    if not ble_device:
+        raise ConfigEntryNotReady(f"Could not find OralB device with address {address}")
+
+    orlb = OralB(ble_device)
+
+    @callback
+    def _async_update_ble(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        orlb.set_ble_device(service_info.device)
+
+    entry.async_on_unload(
+        bluetooth.async_register_callback(
+            hass,
+            _async_update_ble,
+            BluetoothCallbackMatcher({ADDRESS: address}),
+            bluetooth.BluetoothScanningMode.PASSIVE,
+        )
+    )
+
+    async def _async_update():
+        """Update the device state."""
+        await orlb.gatherdata()
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=DOMAIN,
+        update_method=_async_update,
+        update_interval=timedelta(seconds=UPDATE_SECONDS),
+    )
 
-    hass.data.setdefault(DOMAIN, {})
-    # hass.data[DOMAIN][entry.entry_id] = MyApi(...)
+    try:
+        async with async_timeout.timeout(DEVICE_TIMEOUT):
+            await coordinator.async_config_entry_first_refresh()
+    except asyncio.TimeoutError as ex:
+        raise ConfigEntryNotReady(
+            "Unable to communicate with the device; "
+            f"Try moving the Bluetooth adapter closer to {DOMAIN}"
+        ) from ex
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = OralBData(
+        entry.title, orlb, coordinator
```

Comment:
Why do you need to store the entry.title? You can already access the title as `coordinator.config_entry.title`

Reply:
Or even better, since you need to access the oralbdata by entry ID, you probably already have the config entry ?

Reply:
You are right, this is not used anywhere, so I am just going to remove this Title entry.